Login rework: prologue screen

hypest opened the pull request:
As we start to implement the rework of the login flows (aka, the mobile-login project), let's start with implementing the very first screen, the prologue screen, where the user get's asked whether they want to log in or sign up.

This PR does not focus on the visual design side as that will be worked on in a subsequent PR.

At the moment, the screen doesn't actually implement any of the functionality presented to the user. Instead, it shows a Toast message that login/signup is not implemented yet. The actual implementation will be done in a subsequent PR.

There are a few unit tests included that test the navigation logic. Almost all tests are basically testing some trivial things, like: login is not implemented so the state machine needs to still be at the initial state.

All in all, this PR tries to lay the foundation upon which to add the real implementation of the prologue and the rest of the screens. Attention has been put on these features:

* Add a feature flag for the new login flows (`LOGIN_WIZARD_STYLE_ACTIVE`) so we can merge new functionality directly in `develop` and easily follow any developments in `develop`. The flag is set to `true` for the dev (`wasabi`) build only. Production/Alpha/Beta builds have it set to `false` for now.

To test:
* Run the `wasabi` build. Have the app data cleaned up or just uninstall/install
* Notice the new login screen, asking you to either choose to log in or sign up. Tap any of the buttons
* A `... is not implemented yet` Toast comes up

As part of testing this PR, please have a look at the way the fragment (`LogInOrSignUpFragment`) communicates with the parent activity (`LoginActivity`) in a format modular enough so the navigation can be tested with Robolectric (unit) tests.

To run the tests through AndroidStudio, open the `LoginLogicTest.java` file, right-click on the class name and click on "Run LoginLogicTest". If the run fails, make sure the run configuration has `-ea` in the VM Options and that the "Working directory" points to `$MODULE_DIR`.

hypest commented:
Hey @aforcier or @maxme , would any of you have some time for picking this one up? The review basically needs to focus on the unit testing setup and the feature flag. Thanks in advance! 🤗

maxme commented on WordPress/build.gradle:
```diff
@@ -46,15 +46,18 @@ android {
 
         zalpha { // alpha version - enable experimental features
             applicationId "org.wordpress.android"
+            buildConfigField "boolean", "LOGIN_WIZARD_STYLE_ACTIVE", "false"
```
I'm not sure how you want to use this feature flag. If you only want to make this available for the debug version, that's the correct implementation. (A feature branch might be easier to manage).

hypest replied:
The idea is to have the flag "machinery" in place and have the feature merged as soon as possible via small PRs. For now it's only for the development version but will escalate as the feature implementation matures. Does that make sense?

maxme replied:
Makes sense but this is a key feature, and if it breaks during development it will be a PITA for other devs (they can edit and disable the feature flag), that's why I think a feature branch is more adapted for this: if it breaks only you will be affected.

hypest replied:
Good point @maxme . I'll retarget to a feature flag to minimise disruption. Thanks!

maxme commented on WordPress/src/main/AndroidManifest.xml:
```diff
@@ -107,6 +107,11 @@
         </activity>
 
         <activity
+            android:name=".ui.accounts.LoginActivity"
```
When we drop the feature flag, will we drop the `SignInActivity` as well? If not I think we should make that activity name more explicit.

hypest replied:
Yes, `SignInActivity` will be superseded and eventually dropped along with the feature flag itself.

maxme replied:
Ok keeping the `LoginActivity` name is great then.

maxme commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/AppPrefs.java:
```diff
@@ -393,6 +396,18 @@ public static int getThemeImageSizeWidth() {
         }
     }
 
+    // Wizard-style login flow
```
I don't understand why you need to store an activated state here, can we only use the feature flag instead? Or are you planning to change the sign in behavior remotely?

hypest replied:
The idea is to mirror how other flags operate and allow it to be activated on-purpose via a magic URL. The deeplink receiver is not there yet though so, unless we can keep it as is for now, I can add the receiver for completeness, or I can remove the pref and add it in a future PR. WDYT?

maxme replied:
👍 you don't need to add the receiver, I was just wondering why we had this.

hypest commented:
Thanks for picking this up @maxme . I've responded to your comments above, let me know if you have any follow up comments. I'll fix the merge conflicts in the meantime. Thanks!

hypest commented:
I've addressed all the comments so far @maxme so, feel free to take another pass on the PR.

I've also enabled Travis to run the (non-connected) tests automatically. We discussed that over chat and it's a good idea 👍

maxme commented:
Same comment as in https://github.com/wordpress-mobile/WordPress-Android/pull/5848#issuecomment-300063992 - if you merge `develop` in these branch directly, it will be a mess to review and might be a problem if you want to rebase.

maxme commented:
:shipit:
